libsmtutil/CHCSmtLib2Interface.cpp

```diff
@@ -432,3 +432,110 @@ std::optional<smtutil::Expression> CHCSmtLib2Interface::invariantsFromSolverResp
 	return Expression::mkAnd(std::move(definitions));
 }
 #undef precondition
+
+namespace
+{
+
+struct LetBindings
+{
+	using BindingRecord = std::vector<SMTLib2Expression>;
+	std::unordered_map<std::string, BindingRecord> bindings;
+	std::vector<std::string> varNames;
+	std::vector<std::size_t> scopeBounds;
+
+	bool has(std::string const& varName) { return bindings.find(varName) != bindings.end(); }
+
+	SMTLib2Expression& operator[](std::string const& varName)
+	{
+		auto it = bindings.find(varName);
+		solAssert(it != bindings.end());
+		solAssert(!it->second.empty());
+		return it->second.back();
+	}
+
+	void pushScope() { scopeBounds.push_back(varNames.size()); }
+
+	void popScope()
+	{
+		smtAssert(!scopeBounds.empty());
+		auto bound = scopeBounds.back();
+		while (varNames.size() > bound)
+		{
+			auto const& varName = varNames.back();
+			auto it = bindings.find(varName);
+			smtAssert(it != bindings.end());
+			auto& record = it->second;
+			record.pop_back();
+			if (record.empty())
+				bindings.erase(it);
+			varNames.pop_back();
+		}
+		scopeBounds.pop_back();
+	}
+
+	void addBinding(std::string name, SMTLib2Expression expression)
+	{
+		auto it = bindings.find(name);
+		if (it == bindings.end())
+			bindings.insert({name, {std::move(expression)}});
+		else
+			it->second.push_back(std::move(expression));
+		varNames.push_back(std::move(name));
+	}
+};
+
+void inlineLetExpressions(SMTLib2Expression& _expr, LetBindings& _bindings)
+{
+	if (isAtom(_expr))
+	{
+		auto const& atom = asAtom(_expr);
+		if (_bindings.has(atom))
+			_expr = _bindings[atom];
+		return;
+	}
+	auto& subexprs = asSubExpressions(_expr);
+	solAssert(!subexprs.empty());
+	auto const& first = subexprs.at(0);
+	if (isAtom(first) && asAtom(first) == "let")
+	{
+		solAssert(subexprs.size() >= 3);
```

Comment:
```suggestion
		solAssert(subexprs.size() == 3);
```
or why can the rest be ignored, if there can be any more subexpressions?

Reply:
Fixed to assert exactly 3 subexpressions.